jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,236 @@
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphMapping;
+import org.jgrapht.Graphs;
+import org.jgrapht.alg.connectivity.ConnectivityInspector;
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.graph.AsGraphUnion;
+import org.jgrapht.graph.SimpleGraph;
+
+import java.util.*;
+import java.util.function.Supplier;
+
+/**
+ * This is an implementation of the AHU algorithm for detecting an (unweighted) isomorphism between two rooted forests.
+ * Please see <a href="http://mathworld.wolfram.com/GraphIsomorphism.html">mathworld.wolfram.com</a> for a complete
+ * definition of the isomorphism problem for general graphs.
+ *
+ * <p>
+ *     The original algorithm was first presented in "Alfred V. Aho and John E. Hopcroft. 1974.
+ *     The Design and Analysis of Computer Algorithms (1st ed., page 84). Addison-Wesley
+ *     Longman Publishing Co., Inc., Boston, MA, USA."
+ * </p>
+ *
+ * <p>
+ *     This implementation runs in linear time (in the number of vertices of the input forests)
+ *     while using a linear amount of memory.
+ * </p>
+ *
+ * <p>
+ *      For an implementation that supports both unrooted and rooted trees see {@link AHUTreeIsomorphismInspector}.
+ * </p>
+ *
+ * <p>
+ *     Note: This implementation requires the input graphs to have valid vertex suppliers
+ *     (see {@link Graph#getVertexSupplier()}).
+ * </p>
+ *
+ * <p>
+ *     Note: This inspector only returns a single mapping (chosen arbitrarily) rather than all possible mappings.
+ * </p>
+ *
+ * @param <V> the type of the vertices
+ * @param <E> the type of the edges
+ *
+ * @author Alexandru Valeanu
+ */
+public class AHUForestIsomorphismInspector<V, E> implements IsomorphismInspector<V, E> {
+    private final Graph<V, E> forest1;
+    private final Graph<V, E> forest2;
+
+    private final Set<V> roots1;
+    private final Set<V> roots2;
+
+    private boolean computed = false;
+    private IsomorphicGraphMapping<V, E> isomorphicMapping;
+
+    /**
+     * Construct a new AHU rooted forest isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input forests are valid trees.
+     *
+     * @param forest1 the first rooted forest
+     * @param roots1 the roots of the first forest
+     * @param forest2 the second rooted forest
+     * @param roots2 the roots of the second forest
+     * @throws NullPointerException if {@code forest1} is {@code null}
+     * @throws NullPointerException if {@code roots1} is {@code null}
+     * @throws NullPointerException if {@code forest2} is {@code null}
+     * @throws NullPointerException if {@code roots2} is {@code null}
+     * @throws IllegalArgumentException if {@code roots1} is empty
+     * @throws IllegalArgumentException if {@code roots2} is empty
+     * @throws IllegalArgumentException if either {@code roots1} or {@code roots2} contain an invalid vertex
+     */
+    public AHUForestIsomorphismInspector(Graph<V, E> forest1, Set<V> roots1, Graph<V, E> forest2, Set<V> roots2){
```

Comment:
I'm not sure if my previous review comment got lost, but I still think there should be validation here to check the GraphType for self loops and multiple edges.  And the interpretation of edge weight/direction should be specified.

Reply:
I am not sure how to check without doing a lot of work in the constructor. A simple check of GraphType won't be enough as it might not contain self loops even if the GraphType allows it (and vice-versa). The original idea was to add `GraphTests.isTree` or `GraphTests.isSimple` to the constructor but that takes linear time.

Reply:
I believe our current recommendation is to apply the cheap GraphType test even though that might be overconservative; search for isAllowingMultipleEdges in VF2AbstractIsomorphismInspector for example.  (We currently only have a few of these checks but that's because we only recently introduced the GraphType metadata.)

Note that for our standard graph implementations, we do in fact enforce the GraphType constraints (e.g. trying to add a self-loop to a graph that doesn't support it will fail).  So the "vice-versa" case isn't possible for those.

@d-michail @jkinable please correct me if my understanding is wrong here

Reply:
Also, in all the cases where you are repeating the same validation for both trees, move it to a helper function.